### ISSUE TITLE
Impossible to refine away 'uninitialized variable'

### ISSUE THREAD

**Issue by qm3ster** [facebook/flow#4144]
It [seems](https://stackoverflow.com/questions/44447641/refining-uninitialized-variable-in-flow?noredirect=1#comment75923933_44447641) that `uninitialized variable` is a different type than `void`/`typeof undefined`
I **can** make an uninitialized variable of type `?T` or `T|undefined`, but then refinements don't work.

Neither throws an error, however in the below case:

```
let lastNumber: ?number
const output = []
if (lastNumber != null) {
  output.push(lastNumber /* lastNumber is Uncovered code */)
}
```
**OR**
```
let lastNumber: number | void
const output = []
if (lastNumber !== undefined) {
  output.push(lastNumber /* lastNumber is Uncovered code */)
}
```
`flow type-at-pos` into `lastNumber` gives `(unknown)` because `uninitialized variable` (at declaration)

But this has full coverage:
```
let lastNumber: ?number = undefined // Initialized with undefined
const output = []
if (lastNumber != null) {
  output.push(lastNumber /* Fully covered, type-at-pos is number */)
}
```
Can't reproduce at flow.org/try as it's just a coverage issue no error even if I assert `(lastNumber: number)`

**Comment by thecotne** [comment]
### TL;DR; so it's not a bug it's a feature!

---

@qm3ster are you sure there is something wrong with it?

i have tried several versions of code and flow seems to be very smart about variable initialization

```js
const output = []

let lastNumber1: ?number
let lastNumber2: ?number

lastNumber1// void
lastNumber2// void

if (Math.random() === 1) {// this make's lastNumber1 possibly initialized 
  lastNumber1 = 1
}

lastNumber1// ?number
lastNumber2// void

if (lastNumber1) {// ?number
  output.push(lastNumber1)// number
}
if (lastNumber2) {// void
  output.push(lastNumber2)// empty
  // I am not sure but in this code there is no way it's actually number 
  // so probably it's correct in a way
}
```

[flow.org/try](https://flow.org/try/#0MYewdgzgLgBCCuUAOiYF4YG0C6AoXANgKawECG0AcvALYBGRATgIwBcMA-GLQ44STHJUeTAEzsuIvoQpRq9JswD0SmADcQASwAmM4QsaiV6rbtyaAZjAAUAWTJQAFgDpGZMNpA1rASnRoMZj8Ab2MnTQgYGjIAayIAckihOSlmGCQQCAhNOgIATxhNME0oTTICTQAvIm0YXBhBWXleNMDcAF98ZObFY0kDPRSDI1UNHXxLG27UkL7uAYaEZERnFAhHa2mDION53g7zK02mqVFZ0dN6uEQUKFX4deP9XjPjIhokKDyr4wBJGDINBgYBAsAg8EYRBgdFQRRg4UioG0UKcTChEWBIBgAHcyAUSokAcAoPByvlgVI6g1jBAsUhGCA6GRcvioITQIxIcTCmAATi8R0gA)


empty in flow is like never in typescript it is used for Exhaustive Checks

you can read about Exhaustive Checks here

- https://medium.com/@ibosz/advance-flow-type-1-exhaustive-checking-with-empty-type-a02e503cd3a0
- or here for typescript https://basarat.gitbooks.io/typescript/docs/types/never.html#use-case-exhaustive-checks

**Comment by qm3ster** [comment]
It was a coverage issue, and years ago.
Closing.
